<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [ ]:
import pandas as pd
import psycopg2

In [ ]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [14]:
import psycopg2

# Подключение к базе данных PostgreSQL
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)


# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [1]:
# текст запроса
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query_3_1 = '''
SELECT COUNT(*) FROM vacancies;
'''

cursor.execute(query_3_1)

# Получение результата
result = cursor.fetchone()[0]

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результата
print(result)



49197


In [ ]:
# результат запроса

Количество вакансий в нашей базе - 49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [2]:
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query = '''
SELECT COUNT(*) FROM employers;
'''

cursor.execute(query)

# Получение результата
result = cursor.fetchone()[0]

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результата
print(result)


23501


In [ ]:
# результат запроса

Количество работодателей - 23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [20]:
# текст запроса
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query = '''
SELECT COUNT(*) FROM areas;
'''

cursor.execute(query)

# Получение результата
result = cursor.fetchone()[0]

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результата
print(result)


1362


In [ ]:
# результат запроса

Количество регионов - 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [3]:
# текст запроса
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query = '''
SELECT COUNT(*) FROM industries;
'''

cursor.execute(query)

# Получение результата
result = cursor.fetchone()[0]

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результата
print(result)


294


In [ ]:
# результат запроса

Количество сфер деятельности в базе - 294

***

In [ ]:
# выводы по предварительному анализу данных

На основе предварительного анализа данных можно сделать следующие выводы:

1. Количество вакансий: В базе данных содержится значительное количество вакансий, что позволяет проводить анализ рынка труда и выявлять тренды.

2. Количество работодателей: Большое количество работодателей указывает на разнообразие компаний, предлагающих работу, что может быть полезно для анализа конкурентной среды.

3. Количество регионов: Наличие множества регионов позволяет анализировать географическое распределение вакансий и работодателей, что может быть полезно для выявления региональных особенностей рынка труда.

4. Количество сфер деятельности: Разнообразие сфер деятельности позволяет анализировать востребованность различных отраслей и профессий, что может быть полезно для карьерного планирования и анализа рынка труда.

Эти данные могут быть использованы для более глубокого анализа, например, для выявления наиболее востребованных профессий, регионов с наибольшим количеством вакансий, а также для анализа зарплатных предложений в различных сферах деятельности.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [4]:
# текст запроса
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query = '''
SELECT areas.name, COUNT(vacancies.id) AS cnt
FROM vacancies
JOIN areas ON vacancies.area_id = areas.id
GROUP BY areas.name
ORDER BY cnt DESC;
'''

cursor.execute(query)

# Получение результатов
results = cursor.fetchall()

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результатов
for row in results:
    print(f"Регион: {row[0]}, Количество вакансий: {row[1]}")


Регион: Москва, Количество вакансий: 5333
Регион: Санкт-Петербург, Количество вакансий: 2851
Регион: Минск, Количество вакансий: 2112
Регион: Новосибирск, Количество вакансий: 2006
Регион: Алматы, Количество вакансий: 1892
Регион: Екатеринбург, Количество вакансий: 1698
Регион: Нижний Новгород, Количество вакансий: 1670
Регион: Казань, Количество вакансий: 1415
Регион: Краснодар, Количество вакансий: 1301
Регион: Самара, Количество вакансий: 1144
Регион: Ростов-на-Дону, Количество вакансий: 1131
Регион: Воронеж, Количество вакансий: 1063
Регион: Ташкент, Количество вакансий: 851
Регион: Красноярск, Количество вакансий: 847
Регион: Нур-Султан, Количество вакансий: 840
Регион: Челябинск, Количество вакансий: 786
Регион: Пермь, Количество вакансий: 771
Регион: Уфа, Количество вакансий: 767
Регион: Томск, Количество вакансий: 692
Регион: Омск, Количество вакансий: 617
Регион: Саратов, Количество вакансий: 579
Регион: Тюмень, Количество вакансий: 547
Регион: Владивосток, Количество вакансий

In [ ]:
# результат запроса

Пятёрка лидеров по количеству вакансий: Москва, Санкт-Петербург, Минск, Новосибирск, Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [76]:
# текст запроса
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query = '''
SELECT COUNT(*) 
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
'''

cursor.execute(query)

# Получение результата
result = cursor.fetchone()[0]

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результата
print(f"Количество вакансий с заполненной зарплатой: {result}")


Количество вакансий с заполненной зарплатой: 24073


In [75]:
# результат запроса

Количество вакансий с заполненной зарплатой: 24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [5]:
# текст запроса
import psycopg2
import pandas as pd
# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)
# Создание курсора
cursor = connection.cursor()

# Выполнение запроса
query = '''
SELECT 
    ROUND(AVG(salary_from)) AS avg_salary_from,
    ROUND(AVG(salary_to)) AS avg_salary_to
FROM vacancies
WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL;
'''

cursor.execute(query)

# Получение результата
result = cursor.fetchone()

# Закрытие соединения
cursor.close()
connection.close()

# Вывод результата
print(f"Средняя нижняя граница зарплаты: {result[0]}")
print(f"Средняя верхняя граница зарплаты: {result[1]}")


Средняя нижняя граница зарплаты: 71065
Средняя верхняя граница зарплаты: 110537


In [32]:
# результат запроса

- Средняя нижняя граница зарплаты: 71065
- Средняя верхняя граница зарплаты: 110537

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [8]:
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = connection.cursor()

sql_query = '''
SELECT schedule, employment, COUNT(*) AS vacancy_count
FROM vacancies
GROUP BY schedule, employment
ORDER BY vacancy_count DESC;
'''

cur.execute(sql_query)
results = cur.fetchall()

for row in results:
    print(f'{row[0]}, {row[1]} : {row[2]} вакансий')

cur.close()
connection.close()


Полный день, Полная занятость : 35367 вакансий
Удаленная работа, Полная занятость : 7802 вакансий
Гибкий график, Полная занятость : 1593 вакансий
Удаленная работа, Частичная занятость : 1312 вакансий
Сменный график, Полная занятость : 940 вакансий
Полный день, Стажировка : 569 вакансий
Вахтовый метод, Полная занятость : 367 вакансий
Полный день, Частичная занятость : 347 вакансий
Гибкий график, Частичная занятость : 312 вакансий
Полный день, Проектная работа : 141 вакансий
Удаленная работа, Проектная работа : 133 вакансий
Гибкий график, Стажировка : 116 вакансий
Сменный график, Частичная занятость : 101 вакансий
Удаленная работа, Стажировка : 64 вакансий
Гибкий график, Проектная работа : 18 вакансий
Сменный график, Стажировка : 12 вакансий
Вахтовый метод, Проектная работа : 2 вакансий
Сменный график, Проектная работа : 1 вакансий


In [34]:
# результат запроса

- Полный день, Полная занятость : 35367 вакансий
- Удаленная работа, Полная занятость : 7802 вакансий
- Гибкий график, Полная занятость : 1593 вакансий
- Удаленная работа, Частичная занятость : 1312 вакансий
- Сменный график, Полная занятость : 940 вакансий
- Полный день, Стажировка : 569 вакансий
- Вахтовый метод, Полная занятость : 367 вакансий
- Полный день, Частичная занятость : 347 вакансий
- Гибкий график, Частичная занятость : 312 вакансий
- Полный день, Проектная работа : 141 вакансий
- Удаленная работа, Проектная работа : 133 вакансий
- Гибкий график, Стажировка : 116 вакансий
- Сменный график, Частичная занятость : 101 вакансий
- Удаленная работа, Стажировка : 64 вакансий
- Гибкий график, Проектная работа : 18 вакансий
- Сменный график, Стажировка : 12 вакансий
- Вахтовый метод, Проектная работа : 2 вакансий
- Сменный график, Проектная работа : 1 вакансий

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [12]:
# текст запроса
import psycopg2

# Подключение к базе данных
connection = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

# Создание курсора
cur = connection.cursor()

# SQL-запрос для подсчёта количества вакансий по опыту работы
sql_query = '''
SELECT experience, COUNT(*) AS vacancy_count
FROM vacancies
GROUP BY experience
ORDER BY vacancy_count ASC;
'''

# Выполнение запроса
cur.execute(sql_query)
results = cur.fetchall()

# Вывод результата
for row in results:
    print(f'{row[0]} : {row[1]} вакансий')

# Закрытие соединений
cur.close()
connection.close()



Более 6 лет : 1337 вакансий
Нет опыта : 7197 вакансий
От 3 до 6 лет : 14511 вакансий
От 1 года до 3 лет : 26152 вакансий


In [36]:
# результат запроса

- Более 6 лет : 1337 вакансий; 
- Нет опыта : 7197 вакансий; 
- От 3 до 6 лет : 14511 вакансий; 
- От 1 года до 3 лет : 26152 вакансий

***

In [37]:
# выводы по детальному анализу вакансий

Детальный анализ вакансий позволяет сделать ряд полезных выводов, которые помогут лучше понимать рынок труда и потребности работодателей. Вот некоторые ключевые моменты, которые можно выявить при таком анализе:

1. Наиболее распространенные требования к опыту работы. Например, выяснив, что большинство вакансий требуют среднего уровня опыта (например, от 1 до 3 лет), соискатели смогут ориентироваться на развитие именно тех компетенций, которые нужны работодателям.

2. Популярные отрасли и профессии. Анализ позволит определить, какие сферы и специальности пользуются наибольшей популярностью, что полезно как начинающим специалистам, так и профессионалам, планирующим смену направления карьеры.

3. Типы занятости и графики работы. Это важно для понимания общих тенденций рынка: гибкий график становится нормой или, наоборот, преобладают традиционные рабочие режимы?

4. Диапазоны заработной платы. Средние показатели по зарплате позволят увидеть реальную картину доходности профессий и отраслей, помогая оценить адекватность ожиданий кандидатов относительно оплаты труда.

5. Требования к образованию и квалификации. Как часто встречаются вакансии, требующие высшего образования, наличие сертификатов или курсов повышения квалификации? Этот аспект также важен для формирования стратегии развития собственных профессиональных навыков.

Такой анализ полезен как компаниям для оптимизации подбора персонала, так и кандидатам для оценки перспектив своей профессиональной траектории.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [16]:
# текст запроса
import psycopg2

# Подключение к базе данных
conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = '''
SELECT e.name, COUNT(v.id) AS vacancy_count
FROM employers e
JOIN vacancies v ON e.id = v.employer_id
GROUP BY e.name
ORDER BY vacancy_count DESC
LIMIT 5;
'''

cur.execute(sql_query)
results = cur.fetchall()

for i, row in enumerate(results, start=1):
    print(f'{i}-е место: компания "{row[0]}", количество вакансий: {row[1]}')

cur.close()
conn.close()




1-е место: компания "Яндекс", количество вакансий: 1933
2-е место: компания "Ростелеком", количество вакансий: 491
3-е место: компания "Тинькофф", количество вакансий: 444
4-е место: компания "СБЕР", количество вакансий: 428
5-е место: компания "Газпром нефть", количество вакансий: 331


In [ ]:
# результат запроса

- 1-е место: компания "Яндекс", количество вакансий: 1933
- 2-е место: компания "Ростелеком", количество вакансий: 491
- 3-е место: компания "Тинькофф", количество вакансий: 444
- 4-е место: компания "СБЕР", количество вакансий: 428
- 5-е место: компания "Газпром нефть", количество вакансий: 331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [18]:
# текст запроса
import psycopg2

# Подключение к базе данных (убрали повторное подключение)
conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

# Первый запрос: количество работодателей и вакансий по регионам
sql_query_1 = '''
SELECT 
    a.name AS region_name,
    COUNT(DISTINCT e.id) AS employer_count,
    COUNT(v.id) AS vacancy_count
FROM 
    areas a
LEFT JOIN 
    employers e ON a.id = e.area
LEFT JOIN 
    vacancies v ON e.id = v.employer_id
GROUP BY 
    a.name
ORDER BY 
    region_name;
'''

cur.execute(sql_query_1)
results_1 = cur.fetchall()

# Для наглядности можно вывести результаты первого запроса
print("Количество работодателей и вакансий по регионам:")
for row in results_1:
    print(f"Регион: {row[0]}, Работодатели: {row[1]}, Вакансии: {row[2]}")

# Второй запрос: регион с наибольшим количеством работодателей без вакансий
sql_query_2 = '''
WITH region_employer_counts AS (
    SELECT 
        a.name AS region_name,
        COUNT(DISTINCT e.id) AS employer_count
    FROM 
        areas a
    LEFT JOIN 
        employers e ON a.id = e.area
    GROUP BY 
        a.name
),
region_vacancy_counts AS (
    SELECT 
        a.name AS region_name,
        COUNT(v.id) AS vacancy_count
    FROM 
        areas a
    LEFT JOIN 
        vacancies v ON a.id = v.area_id
    GROUP BY 
        a.name
)
SELECT 
    rec.region_name,
    rec.employer_count
FROM 
    region_employer_counts rec
JOIN 
    region_vacancy_counts rvc ON rec.region_name = rvc.region_name
WHERE 
    rvc.vacancy_count = 0
ORDER BY 
    rec.employer_count DESC
LIMIT 1;
'''

cur.execute(sql_query_2)
result_2 = cur.fetchone()

print("nРегион с наибольшим количеством работодателей без вакансий:")
if result_2:
    print(f"Регион: {result_2[0]}, Работодатели: {result_2[1]}")
else:
    print("Таких регионов нет.")

cur.close()
conn.close()



Количество работодателей и вакансий по регионам:
Регион: Абакан, Работодатели: 13, Вакансии: 18
Регион: Абан, Работодатели: 0, Вакансии: 0
Регион: Абатское, Работодатели: 0, Вакансии: 0
Регион: Абдулино, Работодатели: 0, Вакансии: 0
Регион: Абинск, Работодатели: 2, Вакансии: 2
Регион: Абхазия, Работодатели: 0, Вакансии: 0
Регион: Австралия, Работодатели: 6, Вакансии: 13
Регион: Австрия, Работодатели: 1, Вакансии: 4
Регион: Агинское (Забайкальский АО), Работодатели: 0, Вакансии: 0
Регион: Агрыз, Работодатели: 1, Вакансии: 0
Регион: Адамовка, Работодатели: 1, Вакансии: 0
Регион: Адлер, Работодатели: 2, Вакансии: 2
Регион: Адыгейск, Работодатели: 0, Вакансии: 0
Регион: Адыге-Хабль, Работодатели: 0, Вакансии: 0
Регион: Азербайджан, Работодатели: 17, Вакансии: 17
Регион: Азов, Работодатели: 2, Вакансии: 1
Регион: Айхал, Работодатели: 0, Вакансии: 0
Регион: Аксай (Казахстан), Работодатели: 4, Вакансии: 7
Регион: Аксай (Ростовская область), Работодатели: 4, Вакансии: 8
Регион: Актау, Работода

In [ ]:
# результат запроса

Регион с наибольшим количеством работодателей без вакансий:
- Регион: Россия, Работодатели: 410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [20]:
# текст запроса
import psycopg2

# Подключение к базе данных
conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = '''
SELECT 
    e.id AS employer_id,
    e.name AS employer_name,
    COUNT(DISTINCT v.area_id) AS region_count
FROM 
    employers e
JOIN 
    vacancies v ON e.id = v.employer_id
GROUP BY 
    e.id, e.name
ORDER BY 
    region_count DESC;
'''

cur.execute(sql_query)
results = cur.fetchall()

print("Количество регионов для каждого работодателя:")
for row in results:
    print(f"Работодатель ID: {row[0]}, Название: {row[1]}, Количество регионов: {row[2]}")

cur.close()
conn.close()


Количество регионов для каждого работодателя:
Работодатель ID: 1740, Название: Яндекс, Количество регионов: 181
Работодатель ID: 2748, Название: Ростелеком, Количество регионов: 152
Работодатель ID: 5724811, Название: Спецремонт, Количество регионов: 116
Работодатель ID: 5130287, Название: Поляков Денис Иванович, Количество регионов: 88
Работодатель ID: 3682876, Название: ООО ЕФИН , Количество регионов: 71
Работодатель ID: 7944, Название: Совкомбанк, Количество регионов: 63
Работодатель ID: 3776, Название: МТС, Количество регионов: 55
Работодатель ID: 53797, Название: ЭФКО, Управляющая компания, Количество регионов: 49
Работодатель ID: 3776815, Название: КРОН, Количество регионов: 48
Работодатель ID: 4352, Название: Почта России, Количество регионов: 48
Работодатель ID: 622121, Название: MCORE, Количество регионов: 46
Работодатель ID: 197135, Название: ИК СИБИНТЕК, Количество регионов: 46
Работодатель ID: 1473866, Название: Сбербанк-Сервис, Количество регионов: 45
Работодатель ID: 1947

In [ ]:
# результат запроса

Количество регионов для каждого работодателя:
- Работодатель ID: 1740, Название: Яндекс, Количество регионов: 181
- ...

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [22]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = '''
SELECT 
    COUNT(*) AS employer_count
FROM 
    employers e
LEFT JOIN 
    employers_industries ei ON e.id = ei.employer_id
WHERE 
    ei.industry_id IS NULL;
'''

cur.execute(sql_query)
result = cur.fetchone()

print(f"Количество работодателей без сферы деятельности: {result[0]}")

cur.close()
conn.close()


Количество работодателей без сферы деятельности: 8419


In [ ]:
# результат запроса

- Количество работодателей без сферы деятельности: 8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [24]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

# SQL-запрос для получения названия компании
sql_query = '''
WITH employer_industries_count AS (
    SELECT 
        e.id,
        e.name,
        COUNT(ei.industry_id) AS industry_count
    FROM 
        employers e
    LEFT JOIN 
        employers_industries ei ON e.id = ei.employer_id
    GROUP BY 
        e.id, e.name
)
SELECT 
    name
FROM 
    employer_industries_count
WHERE 
    industry_count = 4
ORDER BY 
    name
OFFSET 2
LIMIT 1;
'''

# Выполнение запроса
cur.execute(sql_query)
result = cur.fetchone()

# Вывод результата
print(f"Название компании: {result[0]}")

# Закрытие соединений
cur.close()
conn.close()


Название компании: 2ГИС


In [ ]:
# результат запроса

- Название компании: 2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [26]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

# SQL-запрос для подсчёта количества работодателей
sql_query = '''
SELECT 
    COUNT(DISTINCT e.id) AS employer_count
FROM 
    employers e
JOIN 
    employers_industries ei ON e.id = ei.employer_id
JOIN 
    industries i ON ei.industry_id = i.id
WHERE 
    i.name = 'Разработка программного обеспечения';
'''

# Выполнение запроса
cur.execute(sql_query)
result = cur.fetchone()

# Вывод результата
print(f"Количество работодателей: {result[0]}")

# Закрытие соединений
cur.close()
conn.close()


Количество работодателей: 3553


In [ ]:
# результат запроса

- Количество работодателей: 3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'

response = requests.get(url)
response.encoding = 'utf-8'

soup = BeautifulSoup(response.text, 'html.parser')

tables = soup.find_all('table')
print(f"Найдено таблиц: {len(tables)}")

# Перебираем таблицы, чтобы найти ту, которая содержит нужные заголовки
for i, table in enumerate(tables):
    header = table.find('tr')
    if header:
        headers_text = [th.get_text(strip=True) for th in header.find_all('th')]
        print(f"Таблица {i} заголовки: {headers_text}")

# Предположим, что нужная таблица - это таблица с названием "Город" или похожим
# Например, выберем таблицу с индексом 0 или 1 — в зависимости от вывода выше

# Замените индекс ниже на нужное число после просмотра вывода
target_table = tables[0]

million_cities = []

for row in target_table.find_all('tr')[1:]:
    cells = row.find_all('td')
    if len(cells) > 1:
        city_name = cells[1].get_text(strip=True)
        million_cities.append(city_name)

print("Список городов-миллионников России:")
print(million_cities)

Найдено таблиц: 11
Таблица 0 заголовки: ['№', 'Город', 'Население,тыс. чел.(на 1 января 2025)[2]', 'Население,тыс. чел.(перепись-2021)[3]', 'Приростотносительнопереписи-2010', 'Население,тыс. чел.(перепись-2010)[4]']
Таблица 1 заголовки: ['№', 'Город', 'Население,тыс. чел.(на 1 января 2025)[2]', 'Население,тыс. чел.(перепись-2021)[3]', 'Приростотносительнопереписи-2010', 'Население,тыс. чел.(перепись-2010)[4]']
Таблица 2 заголовки: ['№', 'Город', 'Год достижениячисленности1\xa0млн жителей', 'Население,тыс. чел.(на 1 января 2025)[2]', 'Экономическийрайон', 'Федеральныйокруг']
Таблица 3 заголовки: []
Таблица 4 заголовки: []
Таблица 5 заголовки: ['Численность населения Санкт-Петербурга по годам']
Таблица 6 заголовки: ['1764[5]', '1765[6]', '1780[6]', '1789[6]', '1800[5]', '1811[6]', '1817[6]', '1825[6]', '1833[6]', '1837[6]', '1840[6]']
Таблица 7 заголовки: ['Численность населения Москвы по годам']
Таблица 8 заголовки: ['1350', '1400', '1600', '1638', '1710', '1725', '1738', '1750', '1775

In [35]:
# текст запроса
import psycopg2

million_cities = [
    'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
    'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа',
    'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж',
    'Пермь', 'Волгоград'
]

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = '''
WITH yandex_employers AS (
    SELECT id FROM employers WHERE name = %s
),
vacancies_in_million_cities AS (
    SELECT a.name AS city,
           COUNT(*) AS vacancy_count
    FROM vacancies v
    JOIN yandex_employers ye ON v.employer_id = ye.id
    JOIN areas a ON v.area_id = a.id
    WHERE a.name = ANY(%s)
    GROUP BY a.name
),
total_count AS (
    SELECT 'Total' AS city, COUNT(*) AS vacancy_count
    FROM vacancies v
    JOIN yandex_employers ye ON v.employer_id = ye.id
    JOIN areas a ON v.area_id = a.id
    WHERE a.name = ANY(%s)
)
SELECT city, vacancy_count FROM vacancies_in_million_cities
UNION ALL
SELECT city, vacancy_count FROM total_count
ORDER BY vacancy_count;
'''

cur.execute(sql_query, ('Яндекс', million_cities, million_cities))
results = cur.fetchall()

for city, count in results:
    print(f'{city}: {count} вакансий')
print(f"Всего строк в выборке: {len(results)}")

cur.close()
conn.close()


Омск: 21 вакансий
Челябинск: 22 вакансий
Красноярск: 23 вакансий
Волгоград: 24 вакансий
Пермь: 25 вакансий
Казань: 25 вакансий
Ростов-на-Дону: 25 вакансий
Самара: 26 вакансий
Уфа: 26 вакансий
Краснодар: 30 вакансий
Воронеж: 32 вакансий
Новосибирск: 35 вакансий
Нижний Новгород: 36 вакансий
Екатеринбург: 39 вакансий
Санкт-Петербург: 42 вакансий
Москва: 54 вакансий
Total: 485 вакансий
Всего строк в выборке: 17


In [ ]:
# результат запроса

Список городов-миллионников России:
['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']
- Омск: 21 вакансий
- Челябинск: 22 вакансий
- Красноярск: 23 вакансий
- Волгоград: 24 вакансий
- Пермь: 25 вакансий
- Казань: 25 вакансий
- Ростов-на-Дону: 25 вакансий
- Самара: 26 вакансий
- Уфа: 26 вакансий
- Краснодар: 30 вакансий
- Воронеж: 32 вакансий
- Новосибирск: 35 вакансий
- Нижний Новгород: 36 вакансий
- Екатеринбург: 39 вакансий
- Санкт-Петербург: 42 вакансий
- Москва: 54 вакансий
- Total: 485 вакансий
- Всего строк в выборке: 17

***

In [ ]:
# выводы по анализу работодателей

Выводы по анализу работодателей:

1. Топ работодателей по количеству вакансий показывает, какие компании активно ведут набор сотрудников.

2. Анализ сфер деятельности выявляет наиболее востребованные отрасли, например, IT-сектор с большим числом компаний, специализирующихся на разработке программного обеспечения.

3. География вакансий позволяет определить ключевые регионы с высокой активностью найма, например, миллионники, где представлены вакансии «Яндекса».

4. Разнообразие сфер у работодателей указывает на их профиль — узкоспециализированный или многопрофильный.

5. Распределение вакансий по регионам помогает понять, где сконцентрирован спрос на специалистов.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [2]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = """
SELECT COUNT(*)
FROM vacancies
WHERE LOWER(name) LIKE %s OR LOWER(name) LIKE %s;
"""

cur.execute(sql_query, ('%data%', '%данн%'))
count_related = cur.fetchone()[0]

print(f"Количество вакансий, связанных с данными: {count_related}")

cur.close()
conn.close()


Количество вакансий, связанных с данными: 1771


In [ ]:
# результат запроса

Количество вакансий, связанных с данными: 1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [6]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = """
SELECT COUNT(*)
FROM vacancies
WHERE (
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s OR
    (LOWER(name) LIKE %s AND LOWER(name) NOT LIKE %s) OR
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s
)
AND (
    LOWER(name) LIKE %s OR
    experience = %s OR
    employment = %s
);
"""

params = (
    '%data scientist%',
    '%data science%',
    '%исследователь данных%',
    '%ml%',
    '%html%',
    '%machine learning%',
    '%машинн%обучен%',
    '%junior%',
    'Нет опыта',
    'Стажировка'
)

cur.execute(sql_query, params)
count = cur.fetchone()[0]

print(f"Количество вакансий начинающего дата-сайентиста: {count}")

cur.close()
conn.close()


Количество вакансий начинающего дата-сайентиста: 51


In [ ]:
# результат запроса

Количество вакансий начинающего дата-сайентиста: 51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [10]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = """
SELECT COUNT(*)
FROM vacancies
WHERE (
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s OR
    (LOWER(name) LIKE %s AND LOWER(name) NOT LIKE %s) OR
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s
)
AND (
    LOWER(key_skills) LIKE %s OR
    LOWER(key_skills) LIKE %s
);
"""

params = (
    '%data scientist%',
    '%data science%',
    '%исследователь данных%',
    '%ml%',
    '%html%',
    '%machine learning%',
    '%машинн%обучен%',
    '%sql%',
    '%postgres%'
)

cur.execute(sql_query, params)
count = cur.fetchone()[0]

print(f"Количество вакансий для DS с ключевым навыком SQL или Postgres: {count}")

cur.close()
conn.close()



Количество вакансий для DS с ключевым навыком SQL или Postgres: 229


In [ ]:
# результат запроса

Количество вакансий для DS с ключевым навыком SQL или Postgres: 229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [14]:
# текст запроса
import psycopg2

conn = psycopg2.connect(
    dbname='project_sql',
    user='skillfactory',
    password='cCkxxLVrDE8EbvjueeMedPKt',
    host='84.201.134.129',
    port=5432
)

cur = conn.cursor()

sql_query = """
SELECT COUNT(*)
FROM vacancies
WHERE (
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s OR
    (LOWER(name) LIKE %s AND LOWER(name) NOT LIKE %s) OR
    LOWER(name) LIKE %s OR
    LOWER(name) LIKE %s
)
AND LOWER(key_skills) LIKE %s;
"""

params = (
    '%data scientist%',
    '%data science%',
    '%исследователь данных%',
    '%ml%',
    '%html%',
    '%machine learning%',
    '%машинн%обучен%',
    '%python%'
)

cur.execute(sql_query, params)
count = cur.fetchone()[0]

print(f"Количество вакансий для DS с ключевым навыком Python: {count}")

cur.close()
conn.close()

Количество вакансий для DS с ключевым навыком Python: 357


In [ ]:
# результат запроса

Количество вакансий для DS с ключевым навыком Python: 357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [17]:
# текст запроса
import psycopg2

db_params = {
    'dbname': 'project_sql',
    'user': 'skillfactory',
    'password': 'cCkxxLVrDE8EbvjueeMedPKt',
    'host': '84.201.134.129',
    'port': 5432
}

query = """
SELECT 
    ROUND(AVG(CHAR_LENGTH(key_skills) - CHAR_LENGTH(REPLACE(key_skills, CHR(9), '')) + 1), 2) AS avg_skills
FROM vacancies
WHERE (
        name ILIKE '%data scientist%'
     OR name ILIKE '%data science%'
     OR name ILIKE '%исследователь данных%'
     OR (name LIKE '%ML%' AND name NOT ILIKE '%html%')
     OR name ILIKE '%machine learning%'
     OR name ILIKE '%машинн%обучен%'
      )
  AND key_skills IS NOT NULL
"""

try:
    with psycopg2.connect(**db_params) as conn:
        with conn.cursor() as cur:
            cur.execute(query)
            avg_skills = cur.fetchone()[0]
            print(f"Среднее количество ключевых навыков в вакансиях для DS: {avg_skills}")
except Exception as e:
    print(f"Ошибка при работе с базой данных: {e}")


Среднее количество ключевых навыков в вакансиях для DS: 6.41


In [ ]:
# результат запроса

Среднее количество ключевых навыков в вакансиях для DS: 6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [19]:
# текст запроса
import psycopg2
import pandas as pd

db_params = {
    'dbname': 'project_sql',
    'user': 'skillfactory',
    'password': 'cCkxxLVrDE8EbvjueeMedPKt',
    'host': '84.201.134.129',
    'port': 5432
}

query = """
SELECT experience, 
       ROUND(AVG(
           COALESCE(
               (salary_to + salary_from) / 2,
               salary_to,
               salary_from
           )
       )) AS avg_salary
FROM vacancies
WHERE name ILIKE '%data scientist%'
   OR name ILIKE '%data science%'
   OR name ILIKE '%исследователь данных%'
   OR (name LIKE '%ML%' AND name NOT LIKE '%HTML%')
   OR name ILIKE '%machine learning%'
   OR name ILIKE '%машинн%обучен%'
GROUP BY experience
ORDER BY experience;
"""

try:
    with psycopg2.connect(**db_params) as conn:
        df = pd.read_sql_query(query, conn)
    print(df)
except Exception as e:
    print("Ошибка при выполнении запроса:", e)


C:\Users\admin\AppData\Local\Temp\ipykernel_14044\1678440519.py:35: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, conn)


           experience  avg_salary
0         Более 6 лет         NaN
1           Нет опыта     74643.0
2  От 1 года до 3 лет    139675.0
3       От 3 до 6 лет    243115.0


In [ ]:
# результат запроса

- Более 6 лет - NaN
- Нет опыта - 74643.0
- От 1 года до 3 лет - 139675.0
- От 3 до 6 лет - 243115.0

***

In [ ]:
# выводы по предметному анализу

Исходя из предметного анализа вакансий Data Scientist с разными уровнями опыта и средней зарплатой, можно сделать следующие выводы:

1. Зависимость зарплаты от опыта  
   Средняя зарплата увеличивается с ростом опыта работы. Это логично и ожидаемо — работники с большим опытом и навыками востребованы больше и, как следствие, получают выше.

2. Конкретные уровни опыта и зарплаты (пример из результатов)  
   - Нет опыта — средняя зарплата относительно низкая, но всё же достаточно конкурентная (около 74 000 руб). Это говорит, что работодатели готовы брать новичков на позиции DS с соответствующим уровнем оплаты.  
   - От 1 года до 3 лет — заметный рост зарплаты (примерно 139 000 руб), что подтверждает ценность приобретённого практического опыта.  
   - От 3 до 6 лет — ещё более значительная средняя зарплата (около 240 000 руб), отражающая высокий уровень профессионализма и умений.  
   - Более 6 лет — иногда встречается пустое значение (NaN), возможно из-за малого числа вакансий с явно указанным таким опытом или разброса данных.

3. Рынок для DS с разным опытом  
   Уровень зарплат отражает рыночный спрос и ожидания работодателей. Более высокие зарплаты для опытных специалистов показывают, что работодатели готовы инвестировать в экспертов, в то время как для начинающих ставки остаются доступными для компаний.

4. Рекомендации для соискателей  
   Для начинающих желательно развивать портфолио и набирать опыт, чтобы быстро попасть в область с зарплатами выше среднего. Опыт от 3 лет и выше уже приносит ощутимые бонусы, поэтому цель для карьерного роста — развивать навыки и двигаться к этим уровням.

5. Важность корректных данных  
   Частые пропуски зарплаты или несоответствия в данных (например, NaN для «Более 6 лет») требуют аккуратного анализа и, возможно, дополнительных источников для более полного понимания рынка.

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

### Итоговые выводы:

1. Зависимость зарплат от опыта  
Как и ожидалось, средние зарплаты растут с увеличением опыта работы. Новички получают значительно ниже, тогда как опытные специалисты (от 3 до 6 лет) зарабатывают в среднем в несколько раз больше.

2. Анализ востребованных профессий  
Data Scientist, Machine Learning Engineer и исследователи данных — это востребованные направления с хорошим уровнем оплаты. Учитывая пересечения в названиях вакансий, мы отобрали релевантные позиции для анализа.

3. Пробелы и качество данных  
В данных встречаются пропуски по зарплатам и опыту, а также неоднородность в наименовании вакансий. Это указывает на необходимость учета таких нюансов при построении моделей и анализе, чтобы избежать искажений.

4. Рынок труда и ожидания работодателей  
Работодатели готовы платить больше за опыт и узкоспециализированные навыки. Однако есть спрос и на начинающих соискателей с более низкими зарплатными ожиданиями, что создаёт широкую шкалу возможностей.

---

### Заключение

Проект показал важность опыта при оценке зарплат и подтвердил востребованность профессии Data Scientist. Качественный анализ данных даёт основу для принятия решений как соискателям, так и работодателям. Дальнейшие исследования и модели прогноза позволят глубже понимать рынок труда и эффективно планировать карьерный рост.